stress: failed test in cockroach/storage/storage.test: TestBaseQueuePurgatory

Binary: cockroach/static-tests.tar.gz sha: https://github.com/cockroachdb/cockroach/commits/e23c02052c15dede1efc4ae6b285940814f3dcd5

Stress build found a failed test:

```
=== RUN   TestBaseQueuePurgatory
W160427 07:02:09.402872 gossip/gossip.go:892  not connected to cluster; use --join to specify a connected node
I160427 07:02:09.405212 storage/queue.go:496  [test] purgatory is now empty
--- FAIL: TestBaseQueuePurgatory (0.00s)
	queue_test.go:530: expected purgatory size of 10; got 9
```

Run Details:

```
0 runs so far, 0 failures, over 5s
0 runs so far, 0 failures, over 10s
0 runs so far, 0 failures, over 15s
0 runs so far, 0 failures, over 20s
0 runs so far, 0 failures, over 25s
6 runs completed, 1 failures, over 28s
FAIL
```
Please assign, take a look and update the issue accordingly.

Replicated with `make stress PKG=./storage TESTS=TestBaseQueuePurgatory`:

```
17500 runs so far, 0 failures, over 1m40s

I160705 16:43:13.705669 rand.go:76  Random seed: -8183373358197753265
W160705 16:43:13.707556 gossip/gossip.go:900  not connected to cluster; use --join to specify a connected node
E160705 16:43:13.708347 storage/queue.go:478  [test] (purgatory) on range=10 [/Min-/Min): test error
E160705 16:43:13.708428 storage/queue.go:478  [test] (purgatory) on range=9 [/Min-/Min): test error
E160705 16:43:13.708473 storage/queue.go:478  [test] (purgatory) on range=8 [/Min-/Min): test error
E160705 16:43:13.708521 storage/queue.go:478  [test] (purgatory) on range=7 [/Min-/Min): test error
E160705 16:43:13.708569 storage/queue.go:478  [test] (purgatory) on range=6 [/Min-/Min): test error
E160705 16:43:13.708607 storage/queue.go:478  [test] (purgatory) on range=5 [/Min-/Min): test error
E160705 16:43:13.708652 storage/queue.go:478  [test] (purgatory) on range=4 [/Min-/Min): test error
E160705 16:43:13.708687 storage/queue.go:478  [test] (purgatory) on range=3 [/Min-/Min): test error
E160705 16:43:13.708732 storage/queue.go:478  [test] (purgatory) on range=2 [/Min-/Min): test error
E160705 16:43:13.708873 storage/queue.go:478  [test] (purgatory) on range=5 [/Min-/Min): test error
E160705 16:43:13.708898 storage/queue.go:478  [test] (purgatory) on range=2 [/Min-/Min): test error
E160705 16:43:13.708931 storage/queue.go:478  [test] (purgatory) on range=1 [/Min-/Min): test error
E160705 16:43:13.708963 storage/queue.go:478  [test] (purgatory) on range=10 [/Min-/Min): test error
E160705 16:43:13.708993 storage/queue.go:478  [test] (purgatory) on range=9 [/Min-/Min): test error
E160705 16:43:13.709025 storage/queue.go:478  [test] (purgatory) on range=7 [/Min-/Min): test error
E160705 16:43:13.709056 storage/queue.go:478  [test] (purgatory) on range=6 [/Min-/Min): test error
E160705 16:43:13.709087 storage/queue.go:478  [test] (purgatory) on range=8 [/Min-/Min): test error
E160705 16:43:13.709119 storage/queue.go:478  [test] (purgatory) on range=4 [/Min-/Min): test error
E160705 16:43:13.709149 storage/queue.go:478  [test] (purgatory) on range=3 [/Min-/Min): test error
--- FAIL: TestBaseQueuePurgatory (15.09s)
	queue_test.go:505: expected purgatory size of 10; got 9
	testing.go:113: condition failed to evaluate within 15s: expected 20 processed replicas; got 19
FAIL
```

--- a/storage/queue_test.go
+++ b/storage/queue_test.go
@@ -496,19 +496,21 @@ func TestBaseQueuePurgatory(t *testing.T) {
 		if pc := atomic.LoadInt32(&testQueue.processed); pc != int32(replicaCount) {
 			return errors.Errorf("expected %d processed replicas; got %d", replicaCount, pc)
 		}
+		// We have to loop checking the following conditions because the increment
+		// of testQueue.processed does not happen atomically with the replica being
+		// placed in purgatory.
+		bq.mu.Lock() // Protect access to purgatory and priorityQ.
+		defer bq.mu.Unlock()
+		// Verify that the size of the purgatory map is correct.
+		if l := len(bq.mu.purgatory); l != replicaCount {
+			return errors.Errorf("expected purgatory size of %d; got %d", replicaCount, l)
+		}
+		// ...and priorityQ should be empty.
+		if l := len(bq.mu.priorityQ); l != 0 {
+			return errors.Errorf("expected empty priorityQ; got %d", l)
+		}
 		return nil
 	})
-
-	bq.mu.Lock() // Protect access to purgatory and priorityQ.
-	// Verify that the size of the purgatory map is correct.
-	if l := len(bq.mu.purgatory); l != replicaCount {
-		t.Errorf("expected purgatory size of %d; got %d", replicaCount, l)
-	}
-	// ...and priorityQ should be empty.
-	if l := len(bq.mu.priorityQ); l != 0 {
-		t.Errorf("expected empty priorityQ; got %d", l)
-	}
-	bq.mu.Unlock()
 
 	// Now, signal that purgatoried replicas should retry.
 	testQueue.pChan <- struct{}{}
@@ -517,19 +519,21 @@ func TestBaseQueuePurgatory(t *testing.T) {
 		if pc := atomic.LoadInt32(&testQueue.processed); pc != int32(replicaCount*2) {
 			return errors.Errorf("expected %d processed replicas; got %d", replicaCount*2, pc)
 		}
+		// We have to loop checking the following conditions because the increment
+		// of testQueue.processed does not happen atomically with the replica being
+		// placed in purgatory.
+		bq.mu.Lock() // Protect access to purgatory and priorityQ.
+		defer bq.mu.Unlock()
+		// Verify the replicas are still in purgatory.
+		if l := len(bq.mu.purgatory); l != replicaCount {
+			return errors.Errorf("expected purgatory size of %d; got %d", replicaCount, l)
+		}
+		// ...and priorityQ should be empty.
+		if l := len(bq.mu.priorityQ); l != 0 {
+			return errors.Errorf("expected empty priorityQ; got %d", l)
+		}
 		return nil
 	})
-
-	bq.mu.Lock() // Protect access to purgatory and priorityQ.
-	// Verify the replicas are still in purgatory.
-	if l := len(bq.mu.purgatory); l != replicaCount {
-		t.Errorf("expected purgatory size of %d; got %d", replicaCount, l)
-	}
-	// ...and priorityQ should be empty.
-	if l := len(bq.mu.priorityQ); l != 0 {
-		t.Errorf("expected empty priorityQ; got %d", l)
-	}
-	bq.mu.Unlock()
 
 	// Remove error and reprocess.
 	testQueue.err = nil

